File: src/lib/models/stats.ts

export type StatKey =
    | 'judgement'
    | 'creativity'
    | 'systems'
    | 'precision'
    | 'focus'
    | 'discipline'
    | 'empathy'
    | 'communication'
    | 'resilience'
    | 'adaptability';

export type StatGroup =
    | 'Cognition'
    | 'Perception'
    | 'Drive'
    | 'Social'
    | 'Defense';

export type StatGroupMetadata = {
    name: StatGroup;
    total: number;
    average: number;
};

export type StatDefinition = {
    label: string;
    key: StatKey;
    class: string;
};

// ✅ Extend it cleanly
export type StatWithValue = StatDefinition & {
    value: number;
};

export type GroupedStatDefinition = {
    group: StatGroup;
    items: StatDefinition[];
};

export type GroupedStatSnapshot = {
    group: StatGroupMetadata;
    items: StatWithValue[];
};

export type StatsSnapshot = {
    judgement: number;
    creativity: number;
    systems: number;
    precision: number;
    focus: number;
    discipline: number;
    empathy: number;
    communication: number;
    resilience: number;
    adaptability: number;
};

export const stats_definition:GroupedStatDefinition[] = [
    {
        group: 'Cognition',
        items: [
            { label: 'Judgement', key: 'judgement', class: 'bg-blue-500' },
            { label: 'Creativity', key: 'creativity', class: 'bg-pink-500' }
        ]
    },
    {
        group: 'Perception',
        items: [
            { label: 'Systems Thinking', key: 'systems', class: 'bg-green-500' },
            { label: 'Precision', key: 'precision', class: 'bg-purple-500' }
        ]
    },
    {
        group: 'Drive',
        items: [
            { label: 'Focus', key: 'focus', class: 'bg-blue-500' },
            { label: 'Discipline', key: 'discipline', class: 'bg-amber-500' }
        ]
    },
    {
        group: 'Social',
        items: [
            { label: 'Empathy', key: 'empathy', class: 'bg-rose-500' },
            { label: 'Communication', key: 'communication', class: 'bg-cyan-500' }
        ]
    },
    {
        group: 'Defense',
        items: [
            { label: 'Resilience', key: 'resilience', class: 'bg-emerald-500' },
            { label: 'Adaptability', key: 'adaptability', class: 'bg-indigo-500' }
        ]
    }
];

export function getGroupedStatSnapshot(
    snapshot: StatsSnapshot
): GroupedStatSnapshot[] {
    return stats_definition.map(group => {
        const items = group.items.map(stat => ({
            ...stat,
            value: snapshot[stat.key] ?? 0
        }));

        const total = items.reduce((sum, stat) => sum + stat.value, 0);
        const average = items.length > 0 ? total / items.length : 0;

        return {
            group: {
                name: group.group,
                total,
                average
            },
            items
        };
    });
}

File: src-tauri/src/integrations/snapshots/stats.rs
use crate::sim::person::stats::{StatType, Stats};
use serde::{Deserialize, Serialize};

#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct StatsSnapshot {
    // Cognition
    pub judgement: u16,
    pub creativity: u16,

    // Perception
    pub systems: u16,
    pub precision: u16,

    // Drive
    pub focus: u16,
    pub discipline: u16,

    // Social
    pub empathy: u16,
    pub communication: u16,

    // Defense
    pub resilience: u16,
    pub adaptability: u16,
}

impl From<&Stats> for StatsSnapshot {
    fn from(s: &Stats) -> Self {
        Self {
            judgement: s.get_stat(StatType::Judgement),
            creativity: s.get_stat(StatType::Creativity),
            systems: s.get_stat(StatType::Systems),
            precision: s.get_stat(StatType::Precision),
            focus: s.get_stat(StatType::Focus),
            discipline: s.get_stat(StatType::Discipline),
            empathy: s.get_stat(StatType::Empathy),
            communication: s.get_stat(StatType::Communication),
            resilience: s.get_stat(StatType::Resilience),
            adaptability: s.get_stat(StatType::Adaptability),
        }
    }
}

impl PartialEq<&Stats> for StatsSnapshot {
    fn eq(&self, other: &&Stats) -> bool {
        self.judgement == other.get_stat(StatType::Judgement)
            && self.creativity == other.get_stat(StatType::Creativity)
            && self.systems == other.get_stat(StatType::Systems)
            && self.precision == other.get_stat(StatType::Precision)
            && self.focus == other.get_stat(StatType::Focus)
            && self.discipline == other.get_stat(StatType::Discipline)
            && self.empathy == other.get_stat(StatType::Empathy)
            && self.communication == other.get_stat(StatType::Communication)
            && self.resilience == other.get_stat(StatType::Resilience)
            && self.adaptability == other.get_stat(StatType::Adaptability)
    }
}